common/client/node_fsm_test.go

```diff
@@ -0,0 +1,118 @@
+		n := newTestTransitionNode(t, nil)
+		assert.Equal(t, nodeStateUndialed, n.State())
+		n.setState(nodeStateAlive)
+		assert.Equal(t, nodeStateAlive, n.State())
+		n.setState(nodeStateUndialed)
+		assert.Equal(t, nodeStateUndialed, n.State())
+	})
+
+	t.Run("transitionToAlive", func(t *testing.T) {
+		const destinationState = nodeStateAlive
+		allowedStates := []nodeState{nodeStateDialed, nodeStateInvalidChainID}
+		rpc := newMockNodeClient[types.ID, Head](t)
+		testTransition(t, rpc, testNode.transitionToAlive, destinationState, allowedStates...)
+	})
+
+	t.Run("transitionToInSync", func(t *testing.T) {
+		const destinationState = nodeStateAlive
+		allowedStates := []nodeState{nodeStateOutOfSync}
+		rpc := newMockNodeClient[types.ID, Head](t)
+		testTransition(t, rpc, testNode.transitionToInSync, destinationState, allowedStates...)
+	})
+	t.Run("transitionToOutOfSync", func(t *testing.T) {
+		const destinationState = nodeStateOutOfSync
+		allowedStates := []nodeState{nodeStateAlive}
+		rpc := newMockNodeClient[types.ID, Head](t)
+		rpc.On("DisconnectAll").Once()
+		testTransition(t, rpc, testNode.transitionToOutOfSync, destinationState, allowedStates...)
+	})
+	t.Run("transitionToUnreachable", func(t *testing.T) {
+		const destinationState = nodeStateUnreachable
+		allowedStates := []nodeState{nodeStateUndialed, nodeStateDialed, nodeStateAlive, nodeStateOutOfSync, nodeStateInvalidChainID}
+		rpc := newMockNodeClient[types.ID, Head](t)
+		rpc.On("DisconnectAll").Times(len(allowedStates))
+		testTransition(t, rpc, testNode.transitionToUnreachable, destinationState, allowedStates...)
+	})
+	t.Run("transitionToInvalidChain", func(t *testing.T) {
+		const destinationState = nodeStateInvalidChainID
+		allowedStates := []nodeState{nodeStateDialed, nodeStateOutOfSync}
+		rpc := newMockNodeClient[types.ID, Head](t)
+		rpc.On("DisconnectAll").Times(len(allowedStates))
+		testTransition(t, rpc, testNode.transitionToInvalidChainID, destinationState, allowedStates...)
+	})
+}
+
+func testTransition(t *testing.T, rpc *mockNodeClient[types.ID, Head], transition func(node testNode, fn func()), destinationState nodeState, allowedStates ...nodeState) {
+	node := newTestTransitionNode(t, rpc)
+	for _, allowedState := range allowedStates {
+		m := new(fnMock)
+		node.setState(allowedState)
+		transition(node, m.Fn)
+		assert.Equal(t, destinationState, node.State(), "Expected node to successfully transition from %s to %s state", allowedState, destinationState)
+		m.AssertCalled(t)
+	}
+
+	isAllowed := func(ns nodeState) bool {
+		for _, allowed := range allowedStates {
+			if allowed == ns {
+				return true
+			}
+		}
+
+		return false
+	}
+
+	// noop on attempt to transition from Closed state
+	m := new(fnMock)
+	node.setState(nodeStateClosed)
+	transition(node, m.Fn)
+	m.AssertNotCalled(t)
+	assert.Equal(t, nodeStateClosed, node.State(), "Expected node to remain in closed state on transition attempt")
+
+	for _, nodeState := range allNodeStates {
```

Comment:
Can't you just iterate through the `allowedStates` here instead?

Reply:
No, this block tries to ensure that transition will panic on any not allowed states. 
Block that iterates through `allowedStates` to ensure correct transition is at the beginning of this function.

Reply:
My bad, I wanted to comment on the line below this:
```suggestion
	for _, nodeState := range allNodeStates {
		if slices.Contains(allowedStates, nodeState) || nodeState == nodeStateClosed{
```
This way you can skip the `isAllowed` func